timelineGenerator.js

```diff
@@ -0,0 +1,233 @@
+  HEIGHT = getHeight();
+  GEM_POSITIONS = getGemPositions();
+  BRANCH_POINTS = getBranchPoints();
+  TIME_SCALE = createTimeScale();
+
+  const fakeDom = new JSDOM('<!DOCTYPE html><html><body></body></html>');
+  const body = (await d3).select(fakeDom.window.document).select('body');
+
+  // + 1 is needed or else the bottom axis appears hidden
+  let svg = body
+    .append('svg')
+    .attr('width', WIDTH)
+    .attr('height', HEIGHT + 1);
+
+  svg = addBranchPoints(svg);
+  svg = addGemVersions(svg);
+  svg = addTimeAxis(svg);
+  svg = addGemsAxis(svg);
+
+  fs.writeFileSync(outputPath, body.html());
+};
+
+const setGems = (integratedModelsPath) => {
+  const gems = [];
+
+  const intputDirFiles = fs.readdirSync(integratedModelsPath);
+  for (let i = 0; i < intputDirFiles.length; i++) {
+    const filePath = path.join(integratedModelsPath, intputDirFiles[i]);
+    const stat = fs.lstatSync(filePath);
+    if (stat.isDirectory() && intputDirFiles[i][0] != '.') {
+      const rawJson = fs.readFileSync(`${filePath}/gemRepository.json`, 'utf8');
+      gems.push(JSON.parse(rawJson));
+    }
+  }
+
+  return gems.map((versions) =>
+    versions.filter((v) => v.id.split('.')[2] === '0'),
+  );
+};
+
+const getHeight = () => GEMS.length * 50 + 100;
+
+const getModelFromId = (id) => id.split('-')[0];
+
+const getGemPositions = () =>
+  GEMS.reduce(
+    (acc, [v], i) => ({
+      ...acc,
+      [getModelFromId(v.id)]: HEIGHT - (i + 1) * 50,
+    }),
+    {},
+  );
+
+const getBranchPoints = () => {
+  BRANCH_POINTS = {};
+
+  Array.from(GEMS).forEach(([v]) => {
+    if (v.externalParentId.length > 0) {
+      BRANCH_POINTS[v.id] = v.externalParentId;
+    }
+  });
+
+  return BRANCH_POINTS;
+};
+
+const createTimeScale = () => {
+  const allVersionDates = GEMS.flat().map((v) => new Date(v.releaseDate));
+  const earliestDate = allVersionDates.sort((a, b) => a - b)[0];
+  const timePadding = 60 * 60 * 24 * 30 * 3 * 1000; // 3 months
+  const startDate = new Date(earliestDate - timePadding);
+  const currentDate = new Date(Date.now() + timePadding);
+  return d3.scaleTime().domain([startDate, currentDate]).range([0, WIDTH]);
+};
+
+const addTimeAxis = (svg) => {
+  const xAxis = d3.axisTop().scale(TIME_SCALE);
+  svg
+    .append('g')
+    .attr('transform', `translate(0, ${HEIGHT})`)
+    .style('font-weight', 'bold')
+    .call(xAxis);
+  return svg;
+};
+
+const addGemsAxis = (svg) => {
+  const kvFlippedGemPositions = Object.fromEntries(
+    Object.entries(GEM_POSITIONS).map(([k, v]) => [v, k]),
+  );
+
+  const yScaleLeft = d3.scaleLinear().domain([0, HEIGHT]).range([HEIGHT, 0]);
+  const yAxisLeft = d3
+    .axisRight(yScaleLeft)
+    .tickValues(Object.keys(kvFlippedGemPositions))
+    .tickFormat((x) => kvFlippedGemPositions[x]);
+  svg
+    .append('g')
+    .style('font-size', '12px')
+    .style('font-weight', 'bold')
+    .call(yAxisLeft);
+
+  // const yScaleRight = d3.scaleLinear().domain([0, HEIGHT]).range([HEIGHT, 0]);
```

Comment:
I guess the code below was to also add a right axis? Normally I would suggest to remove commented out code. In this case, maybe it's better to keep it but with a comment, mentioning why this code is not used at the moment (because the resulting right axis was not an improvement at that stage?).

Reply:
Good point! Added an explanation here: https://github.com/MetabolicAtlas/data-generation/pull/27/commits/80fc8846a453e3813edbeb8d0e4cfbc11805635b